src/wifipaf/WiFiPAFEndPoint.cpp

```diff
@@ -0,0 +1,1255 @@
+
+        // Restart message transmission if it was previously paused due to window exhaustion.
+        err = DriveSending();
+        SuccessOrExit(err);
+    }
+
+    // The previous DriveSending() might have generated a piggyback acknowledgement if there was
+    // previously un-acked data.  Otherwise, prepare to send acknowledgement for newly received fragment.
+    //
+    // If local receive window is below immediate ack threshold, AND there is no previous stand-alone ack in
+    // flight, AND there is no pending outbound message fragment on which the ack can and will be piggybacked,
+    // send immediate stand-alone ack to reopen window for sender.
+    //
+    // The "operation in flight" check below covers "pending outbound message fragment" by extension, as when
+    // a message has been passed to the end point via Send(), its next outbound fragment must either be in flight
+    // itself, or awaiting the completion of another in-flight operation.
+    //
+    // If any operation is in flight that is NOT a stand-alone ack, the window size will be checked against
+    // this threshold again when the operation is confirmed.
+    if (mPafTP.HasUnackedData())
+    {
+        if (mLocalReceiveWindowSize <= WIFIPAF_CONFIG_IMMEDIATE_ACK_WINDOW_THRESHOLD &&
+            !mConnStateFlags.Has(ConnectionStateFlag::kOperationInFlight))
+        {
+            ChipLogDebugWiFiPAFEndPoint(WiFiPAF, "sending immediate ack");
+            err = DriveStandAloneAck();
+            SuccessOrExit(err);
+        }
+        else
+        {
+            ChipLogDebugWiFiPAFEndPoint(WiFiPAF, "starting send-ack timer");
+
+            // Send ack when timer expires.
+            err = StartSendAckTimer();
+            SuccessOrExit(err);
+        }
+    }
+
+    // If we've reassembled a whole message...
+    if (mPafTP.RxState() == WiFiPAFTP::kState_Complete)
+    {
+        // Take ownership of message buffer
+        System::PacketBufferHandle full_packet = mPafTP.TakeRxPacket();
+
+        ChipLogDebugWiFiPAFEndPoint(WiFiPAF, "reassembled whole msg, len = %u", static_cast<unsigned>(full_packet->DataLength()));
+
+        // If we have a message received callback, and end point is not closing...
+        if (mWiFiPafLayer != nullptr && mState != kState_Closing)
+        {
+            // Pass received message up the stack.
+            err = mWiFiPafLayer->OnWiFiPAFMsgRxComplete(mSessionInfo, std::move(full_packet));
+        }
+    }
+
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        DoClose(closeFlags, err);
+    }
+
+    return err;
+}
+
+CHIP_ERROR WiFiPAFEndPoint::SendWrite(PacketBufferHandle && buf)
+{
+    mConnStateFlags.Set(ConnectionStateFlag::kOperationInFlight);
+
+    ChipLogDebugBufferWiFiPAFEndPoint(WiFiPAF, buf);
+    Encoding::LittleEndian::Reader reader(buf->Start(), buf->DataLength());
+    DebugPktAckSn(PktDirect_t::kTx, reader, buf->Start());
```

Comment:
I really don't get it.  If you have a reader on the buffer, why are you _also_ manually trying to mess with the buffer pointer?

Reply:
I will change it